src/ServiceControl.TransportAdapter/FailedMessageForwarder.cs

```diff
@@ -15,44 +16,56 @@ class FailedMessageForwarder<TEndpoint, TServiceControl>
     {
         public FailedMessageForwarder(string adapterName, string frontendErrorQueue, string backendErrorQueue, int retryMessageImmeidateRetries, string poisonMessageQueueName,
             Action<TransportExtensions<TEndpoint>> frontendTransportCustomization,
-            Action<TransportExtensions<TServiceControl>> backendTransportCustomization)
+            Action<TransportExtensions<TServiceControl>> backendTransportCustomization,
+            MetricsContext metricsContext)
         {
-            backEndConfig = RawEndpointConfiguration.Create($"{adapterName}.Retry", (context, _) => OnRetryMessage(context), poisonMessageQueueName);
-            backEndConfig.CustomErrorHandlingPolicy(new RetryForwardingFailurePolicy(backendErrorQueue, retryMessageImmeidateRetries, () => retryToAddress));
+            var errorsForwarded = metricsContext.Meter("Errors forwarded", Unit.Custom("Messages"));
+            var errorForwardFailures = metricsContext.Meter("Error forwarding failures", Unit.Custom("Messages"));
+
+            var retriesForwarded = metricsContext.Meter("Retry messages forwarded", Unit.Custom("Messages"));
+            var retryForwardFailures = metricsContext.Meter("Retry message forwarding failures", Unit.Custom("Messages"));
+            var retryReturns = metricsContext.Meter("Retry messages returned to ServiceControl", Unit.Custom("Messages"));
+
+            backEndConfig = RawEndpointConfiguration.Create($"{adapterName}.Retry", (context, _) => OnRetryMessage(context, retriesForwarded), poisonMessageQueueName);
+            backEndConfig.CustomErrorHandlingPolicy(new RetryForwardingFailurePolicy(backendErrorQueue, retryMessageImmeidateRetries, () => retryToAddress, retryForwardFailures, retryReturns));
```

Comment:
Unfortunately `retryToAddres` is only set when starting up so it is not accessible in the constructor. The design problem with NSB.Raw is that the func does not have access to some things that are generated when starting up. I guess the new NSB.Raw will include these things as part of message processing context.